lib/jellyfish/component/hls/manager.ex

```diff
@@ -14,50 +16,60 @@ defmodule Jellyfish.Component.HLS.Manager do
 
   @spec start(Room.id(), pid(), String.t(), map()) :: :ok
   def start(room_id, engine_pid, hls_dir, hls_options) do
-    {:ok, _pid} = GenServer.start(__MODULE__, [room_id, engine_pid, hls_dir, hls_options])
+    {:ok, _pid} =
+      DynamicSupervisor.start_child(
+        Jellyfish.HLS.ManagerSupervisor,
+        {__MODULE__,
+         %{room_id: room_id, engine_pid: engine_pid, hls_dir: hls_dir, hls_options: hls_options}}
+      )
+
     :ok
   end
 
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts)
+  end
+
   @impl true
-  def init([room_id, engine_pid, hls_dir, hls_options]) do
+  def init(%{engine_pid: engine_pid, room_id: room_id} = state) do
     Process.monitor(engine_pid)
-    Logger.info("Initialize s3 uploader")
+    Logger.info("Initialize hls manager, room: #{inspect(room_id)}")
 
-    {:ok, %{room_id: room_id, engine_pid: engine_pid, hls_dir: hls_dir, hls_options: hls_options}}
+    {:ok, state}
   end
 
   @impl true
   def handle_info(
         {:DOWN, _ref, :process, engine_pid, _reason},
-        %{engine_pid: engine_pid} = state
+        %{engine_pid: engine_pid, hls_options: hls_options, hls_dir: hls_dir, room_id: room_id} =
+          state
       ) do
-    unless is_nil(state.hls_options.s3) do
-      state.hls_options.s3
-      |> create_aws_config()
-      |> upload_to_s3(state.hls_dir, state.hls_options.s3.bucket, state.room_id)
-    end
-
-    maybe_remove_hls(state.hls_options, state.hls_dir)
+    unless is_nil(hls_options.s3), do: upload_to_s3(hls_dir, room_id, hls_options.s3)
```

Comment:
Shouldn't that work?
```suggestion
    if hls_options.s3, do: upload_to_s3(hls_dir, room_id, hls_options.s3)
```